### ISSUE TITLE
Compiler error when compiling crates with 1.74.0

### ISSUE THREAD

**Issue by autarch** [rust-lang/rust#118624]
I am seeing a linker error when running `cargo build` with various Rust projects. This seems to be related to upgrading my stable toolchain to 1.74.0. One such project is my [ubi tool](https://github.com/houseabsolute/ubi).

I'm running on Ubuntu 20.04 on x86-64.

See https://gist.github.com/autarch/4243a707d0300be794eb03d2ea65120c for the full error text.

This does not appear to be specific to any particular crate, and I've seen it fail on different crates in different projects. However, if I make a _new_ project and add the crates one at a time then I don't see the error. Instead, the error seems to be triggered after a certain number of crates are compiled (about 15 or so).

I tested with both the latest nightly and with 1.73.0. Neither have this issue, so it seems to be something in 1.74.0.

### Meta

`rustc --version --verbose`:
```
rustc 1.74.0 (79e9716c9 2023-11-13)
binary: rustc
commit-hash: 79e9716c980570bfd1f666e3b16ac583f0168962
commit-date: 2023-11-13
host: x86_64-unknown-linux-gnu
release: 1.74.0
LLVM version: 17.0.4
```

**Comment by aDotInTheVoid** [comment]
I can't reproduce this, could you provide more details?

I tried with:
- ubuntu 20.04 under docker
- x86-64
- rustc 1.74.0 (79e9716c9 2023-11-13) (from rustup)
- cc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
- GNU ld (GNU Binutils for Ubuntu) 2.34
- https://github.com/houseabsolute/ubi/commit/4586a54b2c7d0b98719e0e6388e7da965f4f68f6

**Comment by autarch** [comment]
> I can't reproduce this, could you provide more details?

What additional details would you like? I couldn't find a super simple reproduction recipe. Like I said in the original report, it's not a specific crate triggering it. It's just after a certain number of crates.

Here's some more version info:

```
$> cc --version
cc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0

$> ld --version
GNU ld (GNU Binutils for Ubuntu) 2.34
```

So that's the same as your Docker image.

I'm quite mystified. This seems like it happened out of nowhere. I'm pretty sure I updated to Rust 1.74.0 right when it was released, but I can't recall if I tried using it before yesterday. Of course, it's also possible it's something about my local configuration or other Ubuntu packages I have installed.

**Comment by aDotInTheVoid** [comment]
> What additional details would you like?

Not super sure TBH. The real answer is "whatever it takes to find a reproducable setup for this", but that doesn't help you find it.

Can you reproduce this with a `Dockerfile`? If not, it might be some weird interaction between your environment and the linker. Are you setting anything in RUSTFLAGS / `.cargo/config` / etc?

**Comment by autarch** [comment]
I tried compiling with Rust 1.74.0 in an Ubuntu 20.04 Docker container and it worked fine, so it's definitely something in my local system.

And then ...

I tried upgrading `sccache` (I had an old version, 0.4.1), and that seems to have fixed it. I swear I tried doing some builds with `RUSTC_WRAPPER` unset and still saw this issue, but maybe it was still somehow being used.

I'll close this. Thanks for looking into it!
